flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/operators/join/interval/asyncprocess/AsyncTimeIntervalJoin.java

```diff
@@ -0,0 +1,345 @@
+    private transient ValueState<Long> leftTimerState;
+    // state to record the timer on the right stream. 0 means no timer set
+    private transient ValueState<Long> rightTimerState;
+
+    protected AsyncTimeIntervalJoin(
+            FlinkJoinType joinType,
+            long leftLowerBound,
+            long leftUpperBound,
+            long allowedLateness,
+            long minCleanUpInterval,
+            InternalTypeInfo<RowData> leftType,
+            InternalTypeInfo<RowData> rightType,
+            IntervalJoinFunction joinFunc) {
+        super(
+                joinType,
+                leftLowerBound,
+                leftUpperBound,
+                allowedLateness,
+                minCleanUpInterval,
+                leftType,
+                rightType,
+                joinFunc);
+    }
+
+    @Override
+    public void open(OpenContext openContext) throws Exception {
+        joinFunction.setRuntimeContext(getRuntimeContext());
+        joinFunction.open(DefaultOpenContext.INSTANCE);
+        joinCollector = new EmitAwareCollector();
+
+        // Initialize the data caches.
+        ListTypeInfo<Tuple2<RowData, Boolean>> leftRowListTypeInfo =
+                new ListTypeInfo<>(new TupleTypeInfo<>(leftType, BasicTypeInfo.BOOLEAN_TYPE_INFO));
+        MapStateDescriptor<Long, List<Tuple2<RowData, Boolean>>> leftMapStateDescriptor =
+                new MapStateDescriptor<>(
+                        "IntervalJoinLeftCache", BasicTypeInfo.LONG_TYPE_INFO, leftRowListTypeInfo);
+        leftCache = getRuntimeContext().getMapState(leftMapStateDescriptor);
+
+        ListTypeInfo<Tuple2<RowData, Boolean>> rightRowListTypeInfo =
+                new ListTypeInfo<>(new TupleTypeInfo<>(rightType, BasicTypeInfo.BOOLEAN_TYPE_INFO));
+        MapStateDescriptor<Long, List<Tuple2<RowData, Boolean>>> rightMapStateDescriptor =
+                new MapStateDescriptor<>(
+                        "IntervalJoinRightCache",
+                        BasicTypeInfo.LONG_TYPE_INFO,
+                        rightRowListTypeInfo);
+        rightCache = getRuntimeContext().getMapState(rightMapStateDescriptor);
+
+        // Initialize the timer states.
+        ValueStateDescriptor<Long> leftValueStateDescriptor =
+                new ValueStateDescriptor<>("IntervalJoinLeftTimerState", Long.class);
+        leftTimerState = getRuntimeContext().getState(leftValueStateDescriptor);
+
+        ValueStateDescriptor<Long> rightValueStateDescriptor =
+                new ValueStateDescriptor<>("IntervalJoinRightTimerState", Long.class);
+        rightTimerState = getRuntimeContext().getState(rightValueStateDescriptor);
+
+        paddingUtil = new OuterJoinPaddingUtil(leftType.toRowSize(), rightType.toRowSize());
+    }
+
+    @Override
+    public void onTimer(long timestamp, OnTimerContext ctx, Collector<RowData> out)
+            throws Exception {
+        joinFunction.setJoinKey(ctx.getCurrentKey());
+        joinCollector.setInnerCollector(out);
```

Comment:
the internal state of `joinFunction` and `joinCollector` should be maintained by `DeclaredVariable` since it belongs to the context for async processing.

Reply:
👍 Good suggestion, I rewrote `AsyncTimeIntervalJoin` with `declareChain`.